tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,523 @@
+---
+title: Time, patterns and a little more
+published: 2021-12-02
+tags: haskell, pattern-synonyms, property-testing
+libraries: time
+language: haskell
+author: Felix Miño
+author-name: Felix Miño
+github-profile: felixminom
+description: In this tutorial we’re going to explore the use of pattern synonyms in Haskell, later we’ll give a brief overview to the Stack Builders’ contribution to the `time` Haskell library and finally we’ll review how to use the utilities that were introduced in the contribution.
+---
+
+As Stack Builders one of the core values of our company is contributing to open
+source (OSS). We believe that we have the power to change people’s lives (or at
+least make it easier) by pushing the boundaries of the software industry and
+we’re happy to announce that we were able to contribute to one of Haskell's core
+libraries, [time][time-hackage].
+
+All started with the idea: “What if we bring some utilities that Ruby on Rails
+(RoR) has to Haskell?”. For instance, “How can one get the range of days that a
+month has?” or “How can one get a quarter’s boundaries?” were some of the
+questions that came to our mind. In the following example the RoR code needed to
+get all days in a month will be shown and then Haskell version to do so:
+
+```ruby
+today = Date.today
+today.next_month.all_month
+=> Wed, 01 Nov 2021..Fri, 30 Nov 2021
+```
+
+```haskell
+allMonth :: Day -> [Day]
+allMonth day =
+    let (y, m , _) = toGregorian day
+    in [fromGregorian y m 1 .. fromGregorian y m 31]
+```
+
+and then in `ghci` this function can be invoked:
+
+```bash
+> today <- utctDay <$> getCurrentTime
+> allMonth today
+[2021-11-01 .. 2021-11-30]
+```
+
+As you can see, the Haskell code has to be implemented to latter be called. Why
+the last day of the month is always set to be 31st, if some months have less than
+31 days? Could be crossing your mind right now. That’s because `fromGregorian`
+clips the values to be correct for each month, but now, what is `fromGregorian`?
+Well that’s exactly what this contribution wante to avoid when using the `time`
+library, stop using some functions that really don't know what their purpose
```

Comment:
> that really don't know

Something's missing here (_that don't have a clear purpose_?).

Also, why doesn't `fromGregorian` have a clear purpose? If it doesn't, is the contribution or the `YearMonthDay` the one adding something clearer (I'm only asking about `fromGregorian`)?

Reply:
There are two functions and a pattern to build a Day in the library. Those are `fromGregorian`, `fromJulian` and the `YearMonthDay` pattern. When we started working on this we were using the `fromJulian` function that has some really weird behavior, specially when the Date needs to be clipped. So I wanted to somehow highlight that using this function is not as straightforward as the way it is right now.